src/encoding.js

```diff
@@ -9,43 +9,67 @@
 const assert = require("assert")
 const BN = require("bn.js")
 
-/**
- * Decodes a byte-encoded variable length array of orders. This can be used to
- * decode the result of `BatchExchange.getEncodedUserOrders` and
- * `BatchExchange.getEncodedOrders`.
- * @param {string} bytes The encoded bytes in hex in the form '0x...'
- * @return {Object[]} The decoded array of orders
- */
-function decodeOrders(bytes) {
+class OrderBuffer {
+  constructor(bytes) {
+    this.bytes = bytes
+    this.index = 2 // skip '0x'
+
+    this.readBytes = (size) => this.bytes.slice(this.index, (this.index += size * 2))
+
+    this.decodeAddr = () => `0x${this.readBytes(20)}`
+    this.decodeInt = (size) => new BN(this.readBytes(size / 8), 16).toString()
+
+    return this
+  }
+}
+
+function decodeOrder(bytes) {
+  return {
+    user: bytes.decodeAddr(),
+    sellTokenBalance: bytes.decodeInt(256),
+    buyToken: bytes.decodeInt(16),
+    sellToken: bytes.decodeInt(16),
+    validFrom: bytes.decodeInt(32),
+    validUntil: bytes.decodeInt(32),
+    priceNumerator: bytes.decodeInt(128),
+    priceDenominator: bytes.decodeInt(128),
+    remainingAmount: bytes.decodeInt(128),
+  }
+}
+
+function decodeIndexedOrder(bytes) {
+  return {
+    ...decodeOrder(bytes),
+    orderId: bytes.decodeInt(16),
+  }
+}
```

Comment:
Is this change implemented in this fashion for backwards compatibility?

Reply:
The main motivation was to be able to reuse the existing decoding logic and therefore either append or prepend it to the existing structure.